File: HackerRank/Algorithms/Implementation/3D Surface Area/Solution.java
    static int surfaceArea(int[][] A) {
        int area = 0;
        
        int rows = A.length;
        int columns = A[0].length;
        
        int t, b, u, d, l, r;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                t = b = u = d = l = r = 0;
                
                int v = A[i][j];
                
                if (v > 0) {
                    t++;
                    b++;
                }
                
                // UP
                if (i >= 1) {
                    if (v > A[i-1][j])
                        u = v - A[i-1][j];
                } else
                    u = v;
                
                // DOWN
                if (i < rows - 1) {
                    if (v > A[i+1][j])
                        d = v - A[i+1][j];
                } else
                    d = v;
                
                // LEFT
                if (j >= 1) {
                    if (v > A[i][j-1])
                        l = v - A[i][j-1];
                } else
                    l = v;
                
                // RIGHT
                if (j < columns - 1) {
                    if (v > A[i][j+1])
                        r = v - A[i][j+1];
                } else
                    r = v;
                
                area = area + t + b + u + d + l + r;  
            }
        }
        
        return area;
    }

File: HackerRank/Algorithms/Implementation/3D Surface Area/Solution.py
# https://www.hackerrank.com/challenges/3d-surface-area/problem
# HackerRank: 3D Surface Area
from typing import List


def surface_area(grid: List[List[int]]) -> int:
    """Sum top/bottom faces plus each cell's height exceeding its four neighbors."""
    # Time: O(n)   Space: O(1)
    rows = len(grid)
    cols = len(grid[0])
    area = 0
    for i in range(rows):
        for j in range(cols):
            height = grid[i][j]
            if height == 0:
                continue
            # Top and bottom faces.
            area += 2
            # Four side faces: exposed height versus each neighbor (0 outside grid).
            for ni, nj in ((i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1)):
                neighbor = grid[ni][nj] if 0 <= ni < rows and 0 <= nj < cols else 0
                area += max(0, height - neighbor)
    return area


def test() -> None:
    assert surface_area([[1]]) == 6
    assert surface_area([[1, 3, 4], [2, 2, 3], [1, 2, 4]]) == 60
